### PR TITLE
JDK-8302595: use-after-free related to GraphKit::clone_map

### PR DESCRIPTION
**Pull request by jcking**
`GraphKit::clone_map` duplicates `SafePointNode` and calls `Compile::record_for_igvn`. In some cases `SafePointNode` is not used so `Node::destruct` is called to cleanup. The `Unique_Node_List` returned by `Compile::for_igvn` still references the node which resides in freed memory which may or may not have been reused. We additionally need to remove the node from `Unique_Node_List` as well to prevent this from happening.

I introduced `GraphKit::destruct_map_clone` which undoes `GraphKit::clone_map`. It even clears the type, though I am not sure if this is necessary so feel free to suggest otherwise. Additionally it calls `delete` on `JVMState`, which is a noop, but it seems like the correct thing to do in case its ever changed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8302595](https://bugs.openjdk.org/browse/JDK-8302595): use-after-free related to GraphKit::clone_map

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [3385ae23](https://git.openjdk.org/jdk/pull/12578/files/3385ae23e853219b2c74873d2be0d80c78e2d4ff)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [3385ae23](https://git.openjdk.org/jdk/pull/12578/files/3385ae23e853219b2c74873d2be0d80c78e2d4ff)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12578/head:pull/12578` \
`$ git checkout pull/12578`

Update a local copy of the PR: \
`$ git checkout pull/12578` \
`$ git pull https://git.openjdk.org/jdk pull/12578/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12578`

View PR using the GUI difftool: \
`$ git pr show -t 12578`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12578.diff">https://git.openjdk.org/jdk/pull/12578.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/compile.hpp [hunk 1]
```diff
@@ -937,6 +937,7 @@ class Compile : public Phase {
   PhaseGVN*         initial_gvn()               { return _initial_gvn; }
   Unique_Node_List* for_igvn()                  { return _for_igvn; }
   inline void       record_for_igvn(Node* n);   // Body is after class Unique_Node_List.
+  inline void       remove_for_igvn(Node* n);   // Body is after class Unique_Node_List.
```
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.hpp:940]: May be change comment.

File: src/hotspot/share/opto/graphKit.cpp [hunk 2]
```diff
@@ -735,6 +735,34 @@ SafePointNode* GraphKit::clone_map() {
   return clonemap;
 }
 
+//-----------------------------destruct_map_clone------------------------------
+//
+// Order of destruct is important to increase the likelyhood that memory can be re-used. We need
+// to destruct/free/delete in the exact opposite order as clone_map().
+void GraphKit::destruct_map_clone(SafePointNode* m) {
+  if (m == nullptr) return;
```
Comment (vnkozlov) [inline src/hotspot/share/opto/graphKit.cpp:743]: Please rename `m` to `sfp`. `m` is for `memory`.

Comment (jcking) [reply src/hotspot/share/opto/graphKit.cpp]: Done. Was just following clone_map.

File: src/hotspot/share/opto/compile.hpp (hunk 1 above)
Comment (jcking) [reply src/hotspot/share/opto/compile.hpp]: Updated to say its in node.hpp. Though it is technically still after Unique_Node_List.

**Comment by vnkozlov** [comment]
Looks like there is issue with build - GHA testing build failed.

**Comment by jcking** [comment]
> Looks like there is issue with build - GHA testing build failed.

Looks like it doesn't like setting memory or jvms to `nullptr`. So I'll just avoid clearing them.

**Review by vnkozlov — APPROVED** [review]
Good. I will test it.

File: src/hotspot/share/opto/phaseX.hpp [hunk 3]
```diff
@@ -239,6 +239,10 @@ class PhaseTransform : public Phase {
     assert(t != NULL, "type must not be null");
     _types.map(n->_idx, t);
   }
+  void    clear_type(const Node* n) {
+    if (n->_idx < _types.Size())
+      _types.map(n->_idx, NULL);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.hpp:244]:
```suggestion
    if (n->_idx < _types.Size()) {
      _types.map(n->_idx, NULL);
    }
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by jcking** [comment]
Will wait for @vnkozlov testing before requesting integration.

**Comment by TobiHartmann** [comment]
@vnkozlov's testing all passed.

**Pull request closed without merging**